### ISSUE TITLE
Revert active task definition in case of deploy failure

### ISSUE THREAD

**Issue by StyleT** [fabfuel/ecs-deploy#8]
Currently if you have failed deployment run you'll face with situation when AWS ECS constantly tries to deploy broken version. To prevent this we can automatically, before script exit, switch service to the last working task definition and prevent desperate ECS deployment attempts.

**Comment by haimari** [comment]
i think "The last working task definition" is the problem here.
lets say deployment failed testing, and you roll back. if you just go to revision -1
then try to fix he problem, deploy again, fail again... and roll back you end up with the last non working revision you did before... 

i can't think of a solution for this except from having to store the deployments tests results and this is a bad idea...

I hope you will come up with something
I use this tool in production now and it save me a lot of time
thank you

**Comment by fabfuel** [comment]
The solution would be, to always roll back to the revision which was active before the new deployment started. Just subtracting 1 from the new revision could indeed end in a mess.

PS: Thanks for the 100th ⭐️  @haimari

**Comment by haimari** [comment]
I think that won't be good, because the "LAST ACTIVE" could simply be your last bad build.
it was deployed you it will be the "LAST ACTIVE" however it failed tests and required rollback.

now what happens if your next deployment fails tests too ? you will use the previous active which is not the revision you want.

What i did is:
if the current build tests in my CI passes, i store the entire definition in S3 Bucket
When i roll back i pull it from there.

currently i'm having an issue with this:
    - aws s3 cp s3://${REV_BUCKET}/${CI_PROJECT_NAME} ./
    - REV=`cat ./${CI_PROJECT_NAME}`
    - echo rev is $REV
    - ecs deploy --region ${AWS_REGION} ${CLUSTER_NAME} ${CI_PROJECT_NAME} --task ${REV}

For some reason even though i see the that $REV is correct, it deployed the next revision
I'll find where i'm wrong eventually... 

Hope i explained it well...
Thanks for your help !

**Comment by fabfuel** [comment]
> I think that won't be good, because the "LAST ACTIVE" could simply be your last bad build.
> it was deployed you it will be the "LAST ACTIVE" however it failed tests and required rollback.

If you always deploy with the new `rollback` option, this should not happen.
The only scenario when this could happen is, if you deploy two broken versions at the same time. Because otherwise, `ecs deploy` will always rollback to the last working version after a deployment failure.

Here is the output of the new `rollback` version (still in development):

```
$ ecs deploy my-cluster my-service -t unknown --timeout 10                                                                                                                                                                                                                      07:36:00

Deploying based on task definition: my-service:109

Updating task definition
Changed image of container "nginx" to: "nginx:unknown" (was: "nginx:develop")
Changed image of container "app" to: "app:unknown" (was: "app:develop")

Creating new task definition revision
Successfully created revision: 111

Updating service
Successfully changed task definition to: my-service:111

Deploying new task definition..........
Deployment failed (timeout)

Rolling back to task definition: my-service:109

Updating service
Successfully changed task definition to: my-service:109

Deploying previous task definition..........................
Rollback successful

Deregister task definition revision
Successfully deregistered revision: 111

Deployment failed, but service has been rolled back to previous task definition: my-service:109
```

Best
Fabian

**Comment by haimari** [comment]
Yes, i'll try to explain this better
on some cases (i'm sure this does not effect everyone) you will deploy (replace revision) on your STG and run tests on it **after it has been deployed** ("LAST ACTIVE"). lets say you then run an http request to test functionality and it fails. this build is not good, but as you described from "ecs deploy" perspective that build was **deployed successfully**... however this has nothing to do with the application which is not working well with that build and you probably need to rollback.

This:
```
Deploying new task definition..........
Deployment failed (timeout)
```
Does not make a difference here, this works great as in your example the **task update fails** due to Docker / resources / ALB health check or whatever other deployment failure happened. 

from the application side. the revision considered "GOOD" but it is not...
This is why i keep persistent data (S3) of successful deployments which **also passed testing.**

I agree this should not be something "ecs deploy" command should integrate.
I think it's something the CI developer should deal with.

I now have a complete working CI flow with automatic rollback and it works.
Thanks

**Comment by fabfuel** [comment]
Yes, as you say, your problem should not be solved automatically by a deployment tool. 
I'm glad that you have a working workflow now!

The new rollback mechanism should only solve plain deployment issues. Which means that the new task definition (e.g. a new Docker image version) could not be deployed/the new containers could not be started for whatever reason. `ecs deploy` should not try to solve/rollback on any other application related issues (like failed tests).

Best
Fabian

**Comment by fabfuel** [comment]
Hi everyone, I just released version 1.4.0 which includes the new `--rollback` option:
https://github.com/fabfuel/ecs-deploy/releases/tag/1.4.0

Best
Fabian
